### PR TITLE
FSD fixes for conservation

### PR DESCRIPTION
**Pull request by dabail10**
For detailed information about submitting Pull Requests (PRs) to the CICE-Consortium,
please refer to: <https://github.com/CICE-Consortium/About-Us/wiki/Resource-Index#information-for-developers>

## PR checklist
- [X] Short (1 sentence) summary of your PR: 
This fixes the conservation issues in CICE with the FSD, but also for non-FSD cases!
- [X] Developer(s): 
@dabail10 , @lettie-roach , @cmbitz 
- [X] Suggest PR reviewers from list in the column to the right.
- [X] Please copy the PR test results link or provide a summary of testing completed below.
https://github.com/CICE-Consortium/Test-Results/wiki/icepack_by_hash_forks#925bdea8760a8d833f0550b408226e80a8969378
- How much do the PR code changes differ from the unmodified code? 
    - [ ] bit for bit
    - [X] different at roundoff level
    - [ ] more substantial 
- Does this PR create or have dependencies on CICE or any other models?
    - [X] Yes
    - [ ] No
- Does this PR add any new test cases?
    - [ ] Yes
    - [X] No
- Is the documentation being updated? ("Documentation" includes information on the wiki or in the .rst files from doc/source/, which are used to create the online technical docs at https://readthedocs.org/projects/cice-consortium-cice/.)
    - [ ] Yes
    - [X] No, does the documentation need to be updated at a later time?
        - [ ] Yes
        - [X] No 
- [X] Please document the changes in detail, including _why_ the changes are made.  This will become part of the PR commit log.
This is a bug fix plus some rearranging for conservation in CICE. This impacts FSD cases as well as non-FSD cases. 

There are two main aspects to the bug fix:

1. Initial ice and snow volume values at the beginning of the lateral melt routine are now used for updating the snow and ice enthalpy as well other tracers for lateral melting. This is answer changing in all cases!
2. The other fix is to move the computation of vi0new_lat (lateral growth from FSD) outside of the subroutine fsd_lateral_growth and a check to make sure the category vi0new (vin0new) does not exceed the total vi0new. This is a warning and it recomputes to ensure conservation. This is only answer changing in FSD cases.
3. There is also a change in definition of floe_area_c, so that it is precisely half way between floe_area_l and floe_area_h. The original computation of floe_area_c based on floe_rad_c biases the area towards the lower value.

Update the fsd subroutine arguments (floe_rad_c, floe_rad_l, floe_binwidth, c_fsd_range) to store static data inside Icepack when computed in Icepack rather than pass them back and forth.  Provide an ability to pass out the values from Icepack as optional arguments.

Move the computation of rsiden and get rid of fside which is no longer needed.

Clean up some comments and indentation 

Update swccsm3 test, set ktherm=1

Update interface documentation

There is an associated CICE tag to go with this.

### CODE REVIEW

File: configuration/driver/icedrv_flux.F90
Comment (eclare108213) [inline configuration/driver/icedrv_flux.F90:275]: wouldn't we want to keep the average value as a diagnostic?

File: configuration/driver/icedrv_InitMod.F90
Comment (eclare108213) [inline configuration/driver/icedrv_InitMod.F90:102]: What changes with grid cell in icepack_init_fsd_bounds, that would cause the fsd size arrays to be different?

File: columnphysics/icepack_therm_vertical.F90
Comment (eclare108213) [inline columnphysics/icepack_therm_vertical.F90:2607]: That's correct.  The "Adjust frzmlt" comment here needs to be changed.

File: columnphysics/icepack_therm_vertical.F90
Comment (eclare108213) [inline columnphysics/icepack_therm_vertical.F90:708]: Good catch.  The comments starting with "FYI" are helpful for the review but should be deleted before the PR is merged.

File: columnphysics/icepack_therm_vertical.F90
Comment (eclare108213) [inline columnphysics/icepack_therm_vertical.F90:654]: These comments can be removed before the PR is merged.

File: columnphysics/icepack_therm_vertical.F90
Comment (eclare108213) [inline columnphysics/icepack_therm_vertical.F90:673]: This is inside the "if tr_fsd" block, so could it be deleted?

File: columnphysics/icepack_fsd.F90
Comment (eclare108213) [inline columnphysics/icepack_fsd.F90:568]: moved into add_new_ice routine, so go ahead and delete it here

File: columnphysics/icepack_fsd.F90
Comment (eclare108213) [inline columnphysics/icepack_fsd.F90:591]: moved into add_new_ice routine, so go ahead and delete it here

File: columnphysics/icepack_therm_itd.F90
Comment (eclare108213) [inline columnphysics/icepack_therm_itd.F90:1059]: What's the plan for figuring out if it should be aicen or aicen_init?  Why not use the most recent value?  I guess the vertical melting and lateral melting are proceeding simultaneously, so neither is exactly correct.  (aicen+aicen_init)/2?

File: columnphysics/icepack_therm_itd.F90
Comment (eclare108213) [inline columnphysics/icepack_therm_itd.F90:1123]: Are the changes to _init values bug fixes?

File: columnphysics/icepack_therm_itd.F90
Comment (eclare108213) [inline columnphysics/icepack_therm_itd.F90:1175]: delete comment

File: columnphysics/icepack_therm_itd.F90
Comment (eclare108213) [inline columnphysics/icepack_therm_itd.F90:1637]: delete code (and the comment above it) that's no longer needed

File: columnphysics/icepack_therm_itd.F90
Comment (eclare108213) [reply columnphysics/icepack_therm_itd.F90:1059]: Actually, is aicen_init at the beginning of the timestep or some other point during the timestep?  Seems to be set at the beginning of the lateral_melt subroutine.

File: columnphysics/icepack_therm_itd.F90
Comment (eclare108213) [inline columnphysics/icepack_therm_itd.F90:1001]: If these _init values are not the values at the beginning of the timestep, then I suggest naming them something else to avoid confusion, like vicen0.

**Review by eclare108213 — COMMENTED** [review]
Initial read-through of the code diffs, trying to understand what's happening here and noting a few things that stand out.
What exactly is the conservation bug fix?  There are a lot of changes, so it's not clear.

**Comment by eclare108213** [comment]
GHActions/IcepackTesting failed, so I relaunched it.

**Comment by dabail10** [comment]
One of the failed tests is a Picard convergence. I will do some offline testing.

File: columnphysics/icepack_therm_itd.F90
Comment (dabail10) [reply columnphysics/icepack_therm_itd.F90:1123]: Yes.

File: configuration/driver/icedrv_arrays_column.F90
Comment (dabail10) [reply configuration/driver/icedrv_arrays_column.F90:238]: When updating the driver I thought it would make sense to have different wave fields for each of the four points.

File: configuration/driver/icedrv_arrays_column.F90
Comment (dabail10) [reply configuration/driver/icedrv_arrays_column.F90:257]: Same thing.

File: configuration/driver/icedrv_arrays_column.F90
Comment (dabail10) [reply configuration/driver/icedrv_arrays_column.F90:246]: They don't have to, but I thought it was a good idea.

File: configuration/driver/icedrv_flux.F90
Comment (dabail10) [reply configuration/driver/icedrv_flux.F90:275]: We didn't have rside as a diagnostic before, but we could add it.

File: configuration/driver/icedrv_arrays_column.F90
Comment (dabail10) [reply configuration/driver/icedrv_arrays_column.F90:257]: This one does not need to be different at each point. In fact, this was causing a memory error. I have fixed it.

File: configuration/driver/icedrv_arrays_column.F90
Comment (eclare108213) [reply configuration/driver/icedrv_arrays_column.F90:238]: I really don't think the bin boundaries for the floe size distribution should change from one grid cell to the next.

File: columnphysics/icepack_therm_itd.F90
Comment (eclare108213) [reply columnphysics/icepack_therm_itd.F90:1123]: Are they the only bug fixes?  Please explain this in the main PR info (at the bottom, where it will be copied into the permanent record).

File: configuration/driver/icedrv_arrays_column.F90
Comment (eclare108213) [reply configuration/driver/icedrv_arrays_column.F90:246]: @erinethomas would this change be helpful for E3SM?  If it's not anticipated to be needed in the near future in E3SM or CESM (or other model -- everyone speak up please), I'd prefer to keep the arrays small.

**Review by eclare108213 — COMMENTED** [review]
some responses -

**Comment by apcraig** [comment]
This seems to change answers for all configurations.  Is that expected?

There is also an "out of memory" error for one of the tests.  The derecho test suite, https://github.com/CICE-Consortium/Test-Results/wiki/icepack_by_hash_forks#925bdea8760a8d833f0550b408226e80a8969378, seems to have the same errors thrown up by github actions.

Please add more information to the PR section "Please document the changes in detail, including why the changes are made. This will become part of the PR commit log" listing and explaining what things changed in greater detail.  Also, you indicate a dependency on CICE, does this mean public Icepack interfaces changed and if so, please document that.  Good to also explain why answers changed in a little greater detail.  Thanks.

**Comment by dabail10** [comment]
This test is failing in github actions. This same test is fine on derecho_intel.

### testsuite.macos-latest/conda_macos_restart_col_1x1_swccsm3.macos-latest/logs/icepack.runlog.240730-182013 ###
 
  -------------------------------------
 (picard_solver) picard_solver: Picard solver non-convergence
    (icepack_warnings_setabort) T :file /Users/runner/icepack/columnphysics/icepack_therm_mushy.F90 :line         1399
 (icepack_warnings_aborted) ... (picard_solver)
 (icepack_warnings_aborted) ... (two_stage_solver_nosnow)
 (icepack_warnings_aborted) ... (temperature_changes_salinity)
  (temperature_changes_salinity)temperature_changes_salinity: Picard solver non-convergence (no snow)
 (icepack_warnings_aborted) ... (thermo_vertical)
 (icepack_warnings_aborted) ... (icepack_step_therm1)
  (icepack_step_therm1) ice: Vertical thermo error, cat            1
 (icepack_warnings_aborted) ... (icepack_step_therm1)
 (icepack_warnings_aborted) ... (icepack_step_therm1)
 (icepack_warnings_aborted) ... (icepack_step_therm1)
  
  
 (icedrv_system_abort) ABORTED: 
 (icedrv_system_abort) called from /Users/runner/icepack/configuration/driver/icedrv_step.F90
 (icedrv_system_abort) line number          420
 (icedrv_system_abort) string = (step_therm1)

**Comment by apcraig** [comment]
You could try running that test on derecho with other compilers (gnu, cray?) and also turn on the debug flags to see if it picks anything up.

Could this be a fluke that is related to https://github.com/CICE-Consortium/Icepack/pull/494?

Have you tried running in CICE and does the test suite run fine there?

Are there some modifications that could be impacting this test in particular?

**Comment by dabail10** [comment]
I am able to reproduce it on my macos conda environment. There are NaN's in the mushy-layer solution. This case is using the old shortwave (ccsm3) with ktherm=2. I don't think we should be testing this.

**Comment by apcraig** [comment]
> I am able to reproduce it on my macos conda environment. There are NaN's in the mushy-layer solution. This case is using the old shortwave (ccsm3) with ktherm=2. I don't think we should be testing this.

Can you suggest changes to the swccsm3 test?  Always good to review our test suite and continue to make improvements.  Feel free to also suggest a couple new tests.

**Comment by dabail10** [comment]
I am testing this out with ktherm=1. I think this might be better.

**Comment by dabail10** [comment]
What else is needed here? Can I change the swccsm3 to use ktherm=1?

**Comment by eclare108213** [comment]
> Can I change the swccsm3 to use ktherm=1?

Yes.

I'll look at my previous comments to see what can be marked as resolved.

File: configuration/driver/icedrv_arrays_column.F90
Comment (eclare108213) [reply configuration/driver/icedrv_arrays_column.F90:246]: pinging @erinethomas on this question

File: columnphysics/icepack_therm_itd.F90
Comment (eclare108213) [reply columnphysics/icepack_therm_itd.F90:1001]: The _init notation is okay (we aren't consistent about this in other parts of the code).  I edited the PR language to be more clear.

File: columnphysics/icepack_therm_itd.F90
Comment (eclare108213) [reply columnphysics/icepack_therm_itd.F90:1059]: I would think this should be aicen_init, which isn't defined yet.  Otherwise aicen might be less than puny from the state variable adjustment immediately above, and the fsd wouldn't be updated equivalently.

Comment (dabail10) [reply columnphysics/icepack_therm_itd.F90:1059]: When I look at this code, it seems like the if conditions on rsiden and aicen are kind of redundant. Except in the case where rsiden(n) = 1. This would mean that aicen(n) would go to zero. So, I do think you want a check on the updated aicen(n).

File: configuration/driver/icedrv_arrays_column.F90
Comment (erinethomas) [reply configuration/driver/icedrv_arrays_column.F90:246]: (sorry for delayed response) - the wave frequency space is fixed - so I do not foresee a need to have a grid dependency.... unless we want to go with spatially-varying spectral resolution in WW3 (It IS an idea that has come up .... but one I have no idea how to actually implement in ww3.... )

Comment (erinethomas) [reply configuration/driver/icedrv_arrays_column.F90:246]: if e3sm wishies to go this direction (spatially varying spectral resolution) -  it would take a massive re--write of ww3, and would be a DISTANT future goal (not something I plan on attempting in the near future)

Comment (dabail10) [reply configuration/driver/icedrv_arrays_column.F90:246]: I am confused. We receive a three dimensional field from WW3 which is nx x ny x nfreq. This means we get a different 25 frequencies at every point on the globe. Am I missing something?

**Comment by dabail10** [comment]
Ok. I have cleaned up the driver code and only the changes related to rside / rsiden are now there.

**Comment by dabail10** [comment]
Weird. It is emailing me and telling me gh-actions is failing.

**Comment by apcraig** [comment]
The GH Actions failures are happening with some regularity these days.  Restarting them seems to work.  Not clear what's going on.

File: columnphysics/icepack_therm_itd.F90
Comment (eclare108213) [inline columnphysics/icepack_therm_itd.F90:887]: fix column alignment (picky, sorry, but might as well fix it now)

File: columnphysics/icepack_therm_itd.F90
Comment (eclare108213) [inline columnphysics/icepack_therm_itd.F90:2213]: alignment

File: columnphysics/icepack_therm_itd.F90
Comment (eclare108213) [inline columnphysics/icepack_therm_itd.F90:1139]: alignment could be fixed here too:
```
faero_ocn(k) = faero_ocn(k) &
             + (vsnon_init(n) * (trcrn(nt_aero  +4*(k-1),n)   &
                               + trcrn(nt_aero+1+4*(k-1),n))  &
             +  vicen_init(n) * (trcrn(nt_aero+2+4*(k-1),n)   &
                               + trcrn(nt_aero+3+4*(k-1),n))) &
             * rsiden(n) / dt
```

File: columnphysics/icepack_therm_itd.F90
Comment (eclare108213) [inline columnphysics/icepack_therm_itd.F90:1206]: remove blank line

File: configuration/driver/icedrv_step.F90
Comment (eclare108213) [inline configuration/driver/icedrv_step.F90:116]: It looks like fside is no longer used at all.  Are there any references to it in the docs?

**Review by eclare108213 — COMMENTED** [review]
This PR is very close, just a few clean-up items.

**Comment by dabail10** [comment]
Well, I think I fixed the fsdfixes PR with the latest BGC updates. This was a bit of a mess. Please let me know if you find stuff here.

**Comment by apcraig** [comment]
Sorry about that @dabail10.  Thanks for doing that.  The main thing is that the BGC is merged now and we can move forward.

**Comment by apcraig** [comment]
Can you rerun some tests, both technical and scientific.  The Github Actions is failing with a build error.  Lets just make sure things are still working.  If you want some help from me, just ask.  Thanks!

**Comment by dabail10** [comment]
Found the one syntax error from the BGC merge. My tests are still waiting in the queue on derecho.

**Comment by dabail10** [comment]
Here are the updated test results. The answer changes are due to some reworking of the FSD code and bug fixes which impact non FSD cases.

https://github.com/CICE-Consortium/Test-Results/wiki/icepack_by_hash_forks#b7270c6404e186df30923781dbdd04e9a3cbef6e

**Comment by apcraig** [comment]
Can you create a PR for the CICE changes (minus the update to Icepack).  Have we done a QC test with CICE yet?  Given all answers change, I think we should also run a full test suite on all the compilers on Derecho for Icepack and CICE.  I'm happy to do that whenever it's the appropriate time after further review.

**Comment by dabail10** [comment]
We should do a QC test with CICE I would guess. I am just working on the CICE tag now.

**Comment by eclare108213** [comment]
CICE PR:  https://github.com/CICE-Consortium/CICE/pull/975

**Review by apcraig — CHANGES_REQUESTED** [review]
There need to be checks added somewhere (in icepack_step_therm1 or frzmlt_bottom_lat) that check for present of afsdn, floe_rad_c, and floe_binwidth if tr_fsd is true. Also, it looks like floe_rad_c and floe_binwidth are static fields defined at initialization in Icepack and then passed back and forth between the driver/CICE and Icepack. Maybe we should create some permanent space in fsd to store those variables and stop passing them back and forth. We want to avoid passing in static data over and over, that's not what the interfaces are for.

I think aicen should NOT be an optional argument in frzmlt_bottom_lat. That is not public and is called in one place in Icepack.

There are many places where the alignment could be improved.

Let me know if you need some help with any of these changes.

**Comment by dabail10** [comment]
aicen is only used when tr_fsd is .true. However, you are proposing that afsdn, floe_binwidth, floe_rad_c, and aicen are all brought in via a use statement instead?

**Comment by apcraig** [comment]
> aicen is only used when tr_fsd is .true. However, you are proposing that afsdn, floe_binwidth, floe_rad_c, and aicen are all brought in via a use statement instead?

I'm thinking aicen should be an argument but not optional.  No reason to make it optional as far as I can tell.  That argument is internal to icepack and it will always be passed whether fsd is on or not.  Why make it optional if it'll always be passed and the public interfaces are unaffected?  

In icepack_step_therm1, we could debate whether afsdn should be optional.  I would lean slightly towards saying making it optional (for backwards compatibility) but then it has to be checked whether it's passed when tr_fsd is true.

With floe_rad_c and floe_binwidth, I'm wondering if we should change how those variables are implemented.  Could they live in icepack_fsd such that when fsd is initialized, they are initialized in fsd as module data.  Then we could get rid of those arguments in all the fsd calls.  If the driver/CICE need to know the values of those variables, we could place them in icepack_parameters and add a query routine or we could simply make them in/out in the fsd init routine but still keep a copy in icepack_fsd for use internally.  If those variables are static and identical for all gridcells, then we should store them in Icepack and stop passing them back and forth.

**Comment by dabail10** [comment]
Ok. I am rerunning the testsuite, but I have done a couple things here. I have changed all of the main fsd variables that are not spatially dependent to be module variables in icepack_fsd.F90. I have also made aicen non-optional in the frzmlt_bottom_lateral call. Then I have kept afsdn optional in icepack_step_therm1, but add a check here. Then I allocate l_afsdn when tr_fsd = .true.

**Comment by dabail10** [comment]
Here are the updated test results. 48 tests have different answers, but everything else passed.

https://github.com/CICE-Consortium/Test-Results/wiki/icepack_by_hash_forks#d8c458dda8bafd3409cc56d716b026f0126fa13e

File: configuration/driver/icedrv_init.F90
Comment (apcraig) [inline configuration/driver/icedrv_init.F90:1299]: This is NOT allowed.  If you need floe_rad_c and floe_binwidth in the driver, you need to get it via subroutine arguments, either a query to parameters or add it back into the icepack_init_fsd.  The values can be passed out to the driver AND still live inside Icepack so we don't have to keep passing them in.

**Comment by apcraig** [comment]
Thanks for the recent updates @dabail10.  There are also several comments form @eclare108213 that need to be dealt with.  I just had a quick look and saw a problem in the updates.  If you want me to help clean this up, let me know.  I can create a branch in my fork and then PR back to your branch.

**Comment by dabail10** [comment]
I thought I had addressed all of @eclare108213 's comments above. I guess I need some help prototyping these arrays.

**Comment by eclare108213** [comment]
> I thought I had addressed all of @eclare108213 's comments above.

I resolved the ones that you addressed, and there are still a few left.  Thanks.

**Comment by apcraig** [comment]
No problem, let me create a sandbox with the mods and try to propose something.  It could be I'm misreading the code too.

File: configuration/driver/icedrv_step.F90
Comment (apcraig) [reply configuration/driver/icedrv_step.F90:116]: I think the docs are OK, we'll have to update the interface documentation though, but that is done through automation.

File: configuration/driver/icedrv_init.F90
Comment (apcraig) [reply configuration/driver/icedrv_init.F90:1299]: I'm adding back floe_rad_l, floe_rad_c, floe_binwidth, and c_fsd_range as optional "_out" arguments in icepack_init_fsd_bounds.  Any driver can then have them returned to them as part of that call and store them in the driver.  The fields will always be defined inside Icepack.

**Comment by dabail10** [comment]
> > I thought I had addressed all of @eclare108213 's comments above.
> 
> I resolved the ones that you addressed, and there are still a few left. Thanks.

I am not finding the unresolved comments. I found one about fside / rside as diagnostics. I think I will add these later.

File: columnphysics/icepack_therm_itd.F90
Comment (dabail10) [reply columnphysics/icepack_therm_itd.F90:876]: The diffs were so big with icepack_therm_itd.F90 I did not see your comments!

**Comment by dabail10** [comment]
Hopefully this is ready for review again.

File: columnphysics/icepack_fsd.F90
Comment (eclare108213) [inline columnphysics/icepack_fsd.F90:595]: Should the comment "but ice cannot grow into the area that has grown laterally" be deleted, since the line below it was?

**Review by eclare108213 — CHANGES_REQUESTED** [review]
This is a nice cleanup of the code, plus some bug fixes.  I spotted a few more minor things, reading through the changes.

**Comment by eclare108213** [comment]
This PR also redefines
`floe_area_c = (floe_area_h+floe_area_l)/c2`
instead of using a quadratic formulation.  I'm not sure if that qualifies as a bug fix or simply a different choice in the parameterization, but it should probably be mentioned in the PR commit log (above).

**Comment by apcraig** [comment]
@dabail10, can you review the latest comments and update the code if needed.  I will start testing again too.  Can we get this merged in the next couple days?  Once we merge this, we'll have to update the CICE fsd PR to point to the merged Icepack.  Trying to hold other PRs at the moment, hoping we can finish this up.

**Comment by dabail10** [comment]
I will try to resolve the remaing issues tomorrow.

**Comment by dabail10** [comment]
I have resolved the most recent issues.

**Comment by apcraig** [comment]
I'm running a test suite on derecho.  Unless I hear otherwise, I'll assume the fsd results are as expected.  Will approve and merge when testing is complete, again, unless others provide additional review/feedback.

Are any updates needed in the PR summary?  I will cut and paste that into the merge comment.

**Comment by dabail10** [comment]
> I'm running a test suite on derecho. Unless I hear otherwise, I'll assume the fsd results are as expected. Will approve and merge when testing is complete, again, unless others provide additional review/feedback.
> 
> Are any updates needed in the PR summary? I will cut and paste that into the merge comment.

I did update the PR summary as requested by @eclare108213.

**Comment by apcraig** [comment]
I just updated the PR summary as well.  Are any updates needed in the user guide for the fsd changes?

**Comment by dabail10** [comment]
> I just updated the PR summary as well. Are any updates needed in the user guide for the fsd changes?

Not at this point. This was just a redo of the code. The options for activating it have not changed. I will have a future PR that will be bringing in an alternative wave fracturing method and the documentation will change then.

**Comment by apcraig** [comment]
Testing looks good.

**Comment by apcraig** [comment]
All reviews have been resolved.  Will merge now.  @dabail10, can you update Icepack in the CICE PR to point to the consortium main when you have a chance, then we can merge that too.  Thanks.

**Pull request merged**
